Add support for 'running status'

moseltronics opened the pull request:
Hi rppicomidi, thank you so much for your work. It helps a lot if working with MIDI and USB.

I found a kind of workaround for the lack of dealing with 'running status' which worked for me.

The simple idea is to keep former sent status bytes in order to reuse them and thus creating 'normal' MIDI messages from running status messages.

`if ... stream->index >= stream->total` first helped me avoiding the hangings, so i left it in the code.

Thank you again for your work

moseltronics

Commit:
Add files via upload

Commit:
Update usb_midi_host.c

I tried to cover running status bytes (from old school midi keyboards via UART MIDI).
In order to achieve this the information of stream->buffer[0] and  stream->buffer[1] is reused and the values of the running status bytes are put into  stream->buffer[2] und  stream->buffer[3].
Thus we generate a 'normal' MIDI message to send out to USB.

In the former version there were hangs in the case of running status messages, so i slightly changed the send out routine (stream->index >= stream->total instead of stream->index == stream->total).

Commit:
Cover 'running status' bytes to make old school keyboards work

rppicomidi commented:
@moseltronics Thank you very much for your pull request. It fixes two separate issues.

1. usb_midi_host code fails to correctly encode the CN_ field for system messages. This is a bug. Thank you very much for finding that!
2. usb_midi_host code does not support running status. Supporting it is a feature request. If you have hardware that needs running status support, I agree it needs to be fixed.

The code in the pull request should work if the application calls `tuh_midi_stream_write()` with complete messages in the `buffer` argument. However, if there are multiple MIDI IN ports (for example, ([midi2usbhost-web]https://github.com/rppicomidi/midi2usbhub-web)), it is likely that the application will send incomplete messages on different cables as each byte comes in. I will put specific review comments in the pull request.

rppicomidi commented on usb_midi_host.c:
```diff
@@ -544,90 +544,117 @@ uint32_t tuh_midi_stream_write (uint8_t dev_addr, uint8_t cable_num, uint8_t con
   midi_stream_t *stream = &p_midi_host->stream_write;
 
   uint32_t i = 0;
+  uint8_t const CN_ = cable_num << 4;
```
Good catch!

rppicomidi reviewed (changes requested):
Thank you very much for your pull request. The first comment in the file requires changing the `midih_interface_t` structure so that the `stream_write` field is actually an array. You don't have to do that for me to approve this pull request, but it would be appreciated. What I want is to define a value `CFG_MIDI_HOST_MAX_IN_CABLES` with a default of 16 if it is not defined in `tusb_config.h`. Then in the declaration of `midih_intereface_t`, `midi_stream_t stream_write[CFG_MIDI_HOST_MAX_IN_CABLES];`.

moseltronics commented on usb_midi_host.c:
```diff
@@ -544,90 +544,117 @@ uint32_t tuh_midi_stream_write (uint8_t dev_addr, uint8_t cable_num, uint8_t con
+          stream->buffer[0] = CN_ + MIDI_CIN_SYSCOM_3BYTE;
         }
-        else
+        else        //z.B. MIDI_STATUS_SYSCOM_TUNE_REQUEST
```
Yeah, it was not intended for uploading at first, just wanted to get my device work. I'll change it of course.

moseltronics commented on usb_midi_host.c:
```diff
@@ -544,90 +544,117 @@ uint32_t tuh_midi_stream_write (uint8_t dev_addr, uint8_t cable_num, uint8_t con
+        else stream->buffer[0] = 0;
       }
-      else if ( (msg >= 0x8 && msg <= 0xB) || msg == 0xE )
+      else if ( (msg >= 0x8 && msg < 0xC) || msg == 0xE )
```
Sorry, that was the remnant of numerous code changes. I will restore it to its original state.

Commit:
Removing a bug in my code

There was a bug in my code (in "tuh_midi_stream_write") parsing running status bytes.

It concerns the case of received UART Midi streams (old school) which are to be sent via USB Midi.

This works well now with a 10 year old midi keybard, but there are issues with another 25 year old one, same manufacturer (notes casually stop like an 'all notes off' message).

Thanks rppicomidi for your help.

moseltronics commented on usb_midi_host.c:
```diff
@@ -544,90 +544,117 @@ uint32_t tuh_midi_stream_write (uint8_t dev_addr, uint8_t cable_num, uint8_t con
   midi_stream_t *stream = &p_midi_host->stream_write;
```
Hi rppicomidi,
thanks again for your great work. Indeed i do not have the whole picture of the project, so i don't understand everything you are supposing.
I see the problems which occur if there are two or more streams to be processed. The question is what does the user want. Shall the streams be merged to one (e.g. keyboard and controller) or shall they be routed independently? How can this be controlled? (I am just thinking ...).

rppicomidi replied:
All should be multiplexed onto the single USB endpoint using virtual cables. I will make the changes required for multiple streams after I merge this pull request. Thank you again for your effort and for the bug fixes.

rppicomidi reviewed (approved):
Thanks for this.
